src/persist-client/src/internal/cache.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! In-process caches of [Blob].
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use moka::sync::Cache;
+use mz_ore::bytes::SegmentedBytes;
+use mz_ore::cast::CastFrom;
+use mz_persist::location::{Atomicity, Blob, BlobMetadata, ExternalError};
+use tracing::error;
+
+use crate::cfg::PersistConfig;
+use crate::internal::metrics::Metrics;
+
+// In-memory cache for [Blob].
+#[derive(Debug)]
+pub struct BlobMemCache {
+    metrics: Arc<Metrics>,
+    cache: Cache<String, SegmentedBytes>,
+    blob: Arc<dyn Blob + Send + Sync>,
+}
+
+impl BlobMemCache {
+    pub fn new(
+        cfg: &PersistConfig,
+        metrics: Arc<Metrics>,
+        blob: Arc<dyn Blob + Send + Sync>,
+    ) -> Arc<dyn Blob + Send + Sync> {
+        let cache = Cache::<String, SegmentedBytes>::builder()
+            .max_capacity(u64::cast_from(cfg.blob_cache_mem_limit_bytes))
```

Comment:
thoughts on making this configurable? the limit could be artificially large, and we could dynamically set a multiplier in the weigher fn to play around with it (e.g. max capacity of "1GiB" but set the multiplier to 1GiB/1MiB so it's effectively a 1MiB cache by default)

Reply:
typed this in the review notes, but I'd rather not :D. I'd prefer we do something like https://github.com/MaterializeInc/materialize/pull/19532 before going too much further down the cache road. even by the time we start getting to O(part size), I'd think it'd probably be time to introduce a secondary disk cache layer and the tradeoffs involved just complicate things sooo much. was hoping to keep this PR framed as a really quick win

Reply:
I was more thinking to have the leeway to change it to like, 4MiB if that turns out to make a big difference, but I get that it adds complexity

Reply:
(I'm not sure there's an easy way to model the hit rate with moka's LFU/LRU policies beyond just measuring it empirically)

Reply:
I think my point is that without something like #19532, we won't know if it makes a big difference without just trying things in LaunchDarkly

Reply:
hm, I think I'm not sure why we need to simulate it to that level of fidelity before adding a knob. it seems like the most useful data would be running this on a staging / prod env and just trying a few different sizes and seeing how the numbers look. is the concern for when the sizes get large enough to affect the amount of memory left for the rest of the process?